File: apps/mobile/src/services/stripeService.ts
import { API_URL } from '../config';
import { getFirebaseIdToken } from './brainAuth';
import AsyncStorage from '@react-native-async-storage/async-storage';

const BRAIN_API_URL = API_URL;
const STRIPE_CACHE_KEY = 'ampac_stripe_customer';

// Check if Brain API is reachable
async function isBrainApiAvailable(): Promise<boolean> {
    try {
        const controller = new AbortController();
        const timeoutId = setTimeout(() => controller.abort(), 5000);
        
        // Try the root health endpoint
        const baseUrl = BRAIN_API_URL.replace('/api/v1', '');
        const response = await fetch(`${baseUrl}/health`, {
            method: 'GET',
            signal: controller.signal,
        });
        clearTimeout(timeoutId);
        return response.ok;
    } catch (err) {
        // In dev mode, log the actual error for debugging
        if (__DEV__) {
            console.log('Brain API health check failed:', err);
        }
        return false;
    }
}

export interface StripeCustomer {
    id: string;
    email: string;
    name: string;
    metadata?: Record<string, string>;
    mock?: boolean;
}

export interface StripePaymentIntent {
    id: string;
    customer: string;
    amount: number;
    currency: string;
    status: string;
    client_secret?: string;
    created?: number;
    mock?: boolean;
}

export interface StripeSubscription {
    id: string;
    customer: string;
    items: Array<{ price: string }>;
    status: string;
    current_period_end?: number;
}

export interface StripeCheckoutSession {
    id: string;
    url: string;
}

export const stripeService = {
    createCustomer: async (email: string, name: string, metadata?: Record<string, string>): Promise<StripeCustomer> => {
        // Check cache first
        try {
            const cached = await AsyncStorage.getItem(STRIPE_CACHE_KEY);
            if (cached) {
                const customer = JSON.parse(cached);
                if (customer.email === email) {
                    return customer;
                }
            }
        } catch {}

        // Check if API is available
        const apiAvailable = await isBrainApiAvailable();
        
        if (!apiAvailable) {
            // Return mock customer for offline mode
            console.warn('Brain API unavailable - using offline mode for Stripe');
            const mockCustomer: StripeCustomer = {
                id: `cus_offline_${Date.now()}`,
                email,
                name,
                metadata,
                mock: true,
            };
            await AsyncStorage.setItem(STRIPE_CACHE_KEY, JSON.stringify(mockCustomer));
            return mockCustomer;
        }

        try {
            const token = await getFirebaseIdToken();
            const response = await fetch(`${BRAIN_API_URL}/stripe/customers`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`,
                },
                body: JSON.stringify({ email, name, metadata }),
            });

            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`Failed to create customer: ${response.status} ${errorText}`);
            }

            const customer = await response.json();
            await AsyncStorage.setItem(STRIPE_CACHE_KEY, JSON.stringify(customer));
            return customer;
        } catch (error) {
            console.error('Stripe customer creation error:', error);
            
            // Fallback to mock on error
            const mockCustomer: StripeCustomer = {
                id: `cus_offline_${Date.now()}`,
                email,
                name,
                metadata,
                mock: true,
            };
            await AsyncStorage.setItem(STRIPE_CACHE_KEY, JSON.stringify(mockCustomer));
            return mockCustomer;
        }
    },

    createPaymentIntent: async (customerId: string, amount: number, currency: string = 'usd', metadata?: Record<string, string>): Promise<StripePaymentIntent> => {
        // Check if this is an offline customer
        if (customerId.startsWith('cus_offline_')) {
            return {
                id: `pi_offline_${Date.now()}`,
                customer: customerId,
                amount,
                currency,
                status: 'requires_payment_method',
                client_secret: `pi_offline_${Date.now()}_secret`,
                created: Math.floor(Date.now() / 1000),
                mock: true,
            };
        }

        const apiAvailable = await isBrainApiAvailable();
        if (!apiAvailable) {
            return {
                id: `pi_offline_${Date.now()}`,
                customer: customerId,
                amount,
                currency,
                status: 'requires_payment_method',
                client_secret: `pi_offline_${Date.now()}_secret`,
                created: Math.floor(Date.now() / 1000),
                mock: true,
            };
        }

        try {
            const token = await getFirebaseIdToken();
            const response = await fetch(`${BRAIN_API_URL}/stripe/payment-intents`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`,
                },
                body: JSON.stringify({ customer_id: customerId, amount, currency, metadata }),
            });

            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`Failed to create payment intent: ${response.status} ${errorText}`);
            }

            return await response.json();
        } catch (error) {
            console.error('Stripe payment intent creation error:', error);
            throw error;
        }
    },

    createSubscription: async (customerId: string, priceId: string, metadata?: Record<string, string>): Promise<StripeSubscription> => {
        try {
            const token = await getFirebaseIdToken();
            const response = await fetch(`${BRAIN_API_URL}/stripe/subscriptions`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`,
                },
                body: JSON.stringify({ customer_id: customerId, price_id: priceId, metadata }),
            });

            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`Failed to create subscription: ${response.status} ${errorText}`);
            }

            return await response.json();
        } catch (error) {
            console.error('Stripe subscription creation error:', error);
            throw error;
        }
    },

    createCheckoutSession: async (priceId: string, customerId: string, successUrl: string, cancelUrl: string): Promise<StripeCheckoutSession> => {
        // Handle offline mode
        if (customerId.startsWith('cus_offline_')) {
            throw new Error('Payment processing unavailable offline. Please check your connection.');
        }

        try {
            const token = await getFirebaseIdToken();
            const response = await fetch(`${BRAIN_API_URL}/stripe/checkout-sessions`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`,
                },
                body: JSON.stringify({ price_id: priceId, customer_id: customerId, success_url: successUrl, cancel_url: cancelUrl }),
            });

            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`Failed to create checkout session: ${response.status} ${errorText}`);
            }

            return await response.json();
        } catch (error) {
            console.error('Stripe checkout session creation error:', error);
            throw error;
        }
    },

    listPaymentIntents: async (customerId: string): Promise<StripePaymentIntent[]> => {
        // Return empty for offline customers
        if (customerId.startsWith('cus_offline_')) {
            return [];
        }

        const apiAvailable = await isBrainApiAvailable();
        if (!apiAvailable) {
            return [];
        }

        try {
            const token = await getFirebaseIdToken();
            const response = await fetch(`${BRAIN_API_URL}/stripe/payment-intents/${customerId}`, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': `Bearer ${token}`,
                },
            });

            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`Failed to list payment intents: ${response.status} ${errorText}`);
            }

            return await response.json();
        } catch (error) {
            console.error('Stripe payment intents listing error:', error);
            return [];
        }
    },

    // Check if in offline/mock mode
    isOfflineMode: async (): Promise<boolean> => {
        const cached = await AsyncStorage.getItem(STRIPE_CACHE_KEY);
        if (cached) {
            const customer = JSON.parse(cached);
            return customer.mock === true;
        }
        return !(await isBrainApiAvailable());
    },

    // Clear cached customer (useful for logout)
    clearCache: async (): Promise<void> => {
        await AsyncStorage.removeItem(STRIPE_CACHE_KEY);
    },
};

File: apps/brain/app/services/stripe_service.py
import stripe
from typing import Optional, Dict, Any, List
from app.core.config import get_settings
from app.core.firebase import get_db
from datetime import datetime

settings = get_settings()

class StripeService:
    def __init__(self):
        if settings.STRIPE_SECRET_KEY:
            stripe.api_key = settings.STRIPE_SECRET_KEY
            self.stripe_enabled = True
        else:
            self.stripe_enabled = False
            print("⚠️ Stripe API Key missing. Using Mock Stripe mode.")

    async def create_customer(self, email: str, name: str, metadata: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Creates a new Stripe customer.
        """
        if not self.stripe_enabled:
            # Mock response
            return {
                "id": f"cus_mock_{datetime.now().timestamp()}",
                "email": email,
                "name": name,
                "metadata": metadata or {},
                "mock": True
            }

        try:
            customer = stripe.Customer.create(
                email=email,
                name=name,
                metadata=metadata or {}
            )
            return customer.to_dict()
        except stripe.error.StripeError as e:
            return {"error": str(e)}

    async def create_payment_intent(self, customer_id: str, amount: int, currency: str = "usd", 
                                   metadata: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Creates a payment intent for a customer.
        """
        if not self.stripe_enabled:
            # Mock response
            return {
                "id": f"pi_mock_{datetime.now().timestamp()}",
                "customer": customer_id,
                "amount": amount,
                "currency": currency,
                "status": "requires_payment_method",
                "client_secret": f"pi_mock_{datetime.now().timestamp()}_secret",
                "mock": True
            }

        try:
            intent = stripe.PaymentIntent.create(
                customer=customer_id,
                amount=amount,
                currency=currency,
                metadata=metadata or {},
                automatic_payment_methods={
                    'enabled': True,
                },
            )
            return intent.to_dict()
        except stripe.error.StripeError as e:
            return {"error": str(e)}

    async def create_subscription(self, customer_id: str, price_id: str, 
                                 metadata: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Creates a subscription for a customer.
        """
        if not self.stripe_enabled:
            # Mock response
            return {
                "id": f"sub_mock_{datetime.now().timestamp()}",
                "customer": customer_id,
                "items": [{"price": price_id}],
                "status": "active",
                "current_period_end": int(datetime.now().timestamp()) + 30 * 24 * 60 * 60,
                "mock": True
            }

        try:
            subscription = stripe.Subscription.create(
                customer=customer_id,
                items=[{"price": price_id}],
                metadata=metadata or {}
            )
            return subscription.to_dict()
        except stripe.error.StripeError as e:
            return {"error": str(e)}

    async def list_payment_intents(self, customer_id: str, limit: int = 10) -> List[Dict[str, Any]]:
        """
        Lists payment intents for a customer.
        """
        if not self.stripe_enabled:
            # Mock response
            return [
                {
                    "id": f"pi_mock_{datetime.now().timestamp()}",
                    "customer": customer_id,
                    "amount": 1000,
                    "currency": "usd",
                    "status": "succeeded",
                    "created": int(datetime.now().timestamp()),
                    "mock": True
                }
            ]

        try:
            intents = stripe.PaymentIntent.list(
                customer=customer_id,
                limit=limit
            )
            return [intent.to_dict() for intent in intents.data]
        except stripe.error.StripeError as e:
            return [{"error": str(e)}]

    async def sync_payment_to_firestore(self, payment_data: Dict[str, Any]) -> str:
        """
        Syncs payment data to Firestore for admin dashboard.
        """
        try:
            db = get_db()
            doc_ref = db.collection("payments").document()
            
            payment_record = {
                "customerId": payment_data.get("customer", ""),
                "customerName": payment_data.get("metadata", {}).get("customer_name", "Unknown"),
                "amount": payment_data.get("amount", 0) / 100,  # Convert cents to dollars
                "currency": payment_data.get("currency", "USD"),
                "status": payment_data.get("status", "pending"),
                "createdAt": datetime.utcnow(),
                "invoiceId": payment_data.get("id", ""),
                "subscriptionId": payment_data.get("subscription", ""),
                "stripeData": payment_data
            }
            
            doc_ref.set(payment_record)
            return doc_ref.id
        except Exception as e:
            print(f"Error syncing payment to Firestore: {e}")
            return ""

    async def create_checkout_session(self, price_id: str, customer_id: str, 
                                     success_url: str, cancel_url: str) -> Dict[str, Any]:
        """
        Creates a Stripe checkout session.
        """
        if not self.stripe_enabled:
            # Mock response
            return {
                "id": f"cs_mock_{datetime.now().timestamp()}",
                "url": f"https://checkout.stripe.com/mock/{datetime.now().timestamp()}",
                "mock": True
            }

        try:
            session = stripe.checkout.Session.create(
                payment_method_types=['card'],
                line_items=[{
                    'price': price_id,
                    'quantity': 1,
                }],
                mode='payment',
                customer=customer_id,
                success_url=success_url,
                cancel_url=cancel_url,
            )
            return session.to_dict()
        except stripe.error.StripeError as e:
            return {"error": str(e)}

stripe_service = StripeService()